pkg/main/router.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"strings"
+	"sync"
+)
+
+type Router interface {
+	Delegate(input map[string]interface{}) (*Response, error)
+}
+
+type RouterImpl struct {
+	servers []Server
+	mutex   *sync.Mutex
+}
+
+func NewRouter(servers []Server) *RouterImpl {
+	return &RouterImpl{
+		servers: servers,
+		mutex:   &sync.Mutex{},
+	}
+}
+
+func (r *RouterImpl) Delegate(input map[string]interface{}) (*Response, error) {
+	server, _ := r.findFreeServer()
+	defer r.releaseServer(server)
+
+	var e Error
+	resp, err := server.Invoke(input)
+	if resp != nil {
+		defer resp.Close()
+	}
+	if err != nil {
+		if _, ok := err.(TimeoutError); ok {
+			e = Error{
+				ErrorType: FUNCTION_ERROR,
+				Message:   "Invocation exceeded the timeout",
+			}
+		} else if _, ok := err.(BadRequestError); ok {
+			e = Error{
+				ErrorType: INPUT_ERROR,
+				Message:   "Input invalid",
+			}
+		} else if _, ok := err.(InvocationError); ok {
+			e = Error{
+				ErrorType: FUNCTION_ERROR,
+				Message:   "Failed invoking function.",
+			}
+		} else {
+			e = Error{
+				ErrorType: SYSTEM_ERROR,
+				Message:   "Something went wrong.",
+			}
+		}
+	}
+
+	logs := Logs{
+		Stdout: splitLogsOnNewline(server.Stdout()),
+		Stderr: splitLogsOnNewline(server.Stderr()),
+	}
+
+	context := Context{
+		Logs:  logs,
+		Error: e,
+	}
+
+	respBuf := new(bytes.Buffer)
+	if resp != nil {
+		respBuf.ReadFrom(resp)
+	}
+	respPayload := make(map[string]interface{})
+	json.Unmarshal(respBuf.Bytes(), &respPayload)
+
+	response := &Response{
+		Context: context,
+		Payload: respPayload,
+	}
+
+	return response, nil
+}
+
+func (r *RouterImpl) findFreeServer() (Server, error) {
+	var ret Server
+	for _, server := range servers {
+		r.mutex.Lock()
+		if server.IsIdle() {
```

Comment:
defer r.mutex.Unlock()

Reply:
I think you could probably use sync.Map rather than the locks.

Reply:
Or... do a little research on work pools for golang.  If there is no available server, this is going to spin.

Reply:
I agree that there are currency issues here, but as written it shouldn't block while there are no available servers. Perhaps worse, it fails to process anything when a request comes in and no servers are available. I admit when I first wrote this I was happy to assume that riff and kubeless would be handling the concurrency and wouldn't send more requests than configured servers. That certainly is a recipe for bugs later. Unfortunately even with a worker pool I think I would still need the mutex around my checks to see which of the configured servers is idle/busy because I'm not spawning functionally identical goroutines to do the processing.

Reply:
Is the most recent change kind of what you had in mind?